### PR TITLE
Introducing Reindexer CLI

### LINKED ISSUE #395: DB3 bag is irrecoverable if metadata.yaml doesn't exist
## Description
ros2 db3 bags fails to load or get info on if the metadata yaml is missing

### Expected Behavior
Expecting it to have the ability to generate it offline in case of any corruption during saving the bag

### Actual Behavior
If accompanying metadata yaml doesn't exist with the db3, it fails to load/get info on.
[ERROR] [1588037968.719797386] [rosbag2_storage]: Could not open '/root/Downloads/jobq_bags.db3' with 'sqlite3'. Error: Failed to setup storage. Error: Error when processing SQL statement. SQLite error (11): database disk image is malformed

## To Reproduce
** Steps to reproduce the behavior, e.g.
1. Use any ros2 bag and rename metadata yaml to tmp
2. Run ros2 bag info <bag-name> 

## System (please complete the following information)
 - OS: Ubuntu Bionic
 - ROS 2 Distro: dashing
 - Version: 

## Additional context
Would love to find a way to recover the bag as its 50 GB worth of data.

### PR DESCRIPTION
**Pull request by jhdcs**
Fixes #395 
Followup to #641

This PR adds the ros2bag/rosbag2_py CLI so that end users can reindex their bag files.

### CODE REVIEW

**Comment by jhdcs** [comment]
Oh crud! No, I didn't intend to re-upload the broken test... Also, I might have had something funny go on with the rebasing I did. Looks like you found a place that might have occurred? I'll get your "nitpicks" fixed! (I quoted Nitpicks, because I don't think they're very nit-picky)

File: rosbag2_py/test/test_reindexer.py [hunk 1]
```diff
@@ -43,8 +43,7 @@ def test_reindexer_multiple_files():
     reindexer = rosbag2_py.Reindexer()
     reindexer.reindex(storage_options)
 
-    result_path = Path(__file__).parent.parent / 'resources' / 'reindex_test_bags' \
-        / 'multiple_files' / 'metadata.yaml'
+    result_path = bag_path / 'metadata.yaml'
```
Comment (emersonknapp) [inline rosbag2_py/test/test_reindexer.py]: bag_path is a string because of your earlier cast

File: ros2bag/ros2bag/verb/reindex.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#
+# DISTRIBUTION A. Approved for public release; distribution unlimited.
+# OPSEC #4584.
+#
+# Delivered to the U.S. Government with Unlimited Rights, as defined in DFARS
+# Part 252.227-7013 or 7014 (Feb 2014).
+#
+# This notice must appear in all copies of this file and its derivatives.
+
+# from argparse import FileType
```
Comment (emersonknapp) [inline ros2bag/ros2bag/verb/reindex.py:23]: didn't notice these the first time - can you remove all these commented-out import lines?

**Comment by emersonknapp** [comment]
Sorry - this'll need a quick rebase and conflict resolution - we've been touching some of the same files. It should be a simple resolution, though, we both added a python module to rosbag2_py

**Comment by jhdcs** [comment]
It looks like the rebase was successful on my end, so hopefully this works for you too?

**Comment by emersonknapp** [comment]
Ah - I'm realizing why those database files are re-added, and it's that we moved them to `rosbag2_tests` in the previous step, so that we could be sure to have access to `rosbag2_storage_default_plugins` in `rosbag2_cpp`. As this stands, we do have access to those plugins from `rosbag2_py` as a `<test_depend>`, so i'll just kick off CI now

**Comment by jhdcs** [comment]
Wait, I'm confused. Should I remove the database files? I thought I had stopped re-introducing them. Unless you meant the ones in `rosbag2_py`, which I'm just using as a test. If there's a way to just use one set of database files across all packages that would be nice... Or did I already do that before?

I think I'm making myself confused...

**Comment by emersonknapp** [comment]
Gist: https://gist.githubusercontent.com/emersonknapp/a0f471c5073bd874049e9c10ddb2bfe9/raw/5eae459c861e4b2e910b1530f57b3ca917adc8c2/ros2.repos
BUILD args: --packages-up-to rosbag2_py ros2bag rosbag2_tests
TEST args: --packages-select rosbag2_py ros2bag rosbag2_tests
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/8072
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14188/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8986/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11870/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14290/)

**Comment by emersonknapp** [comment]
These files look like a duplicate of the ones that you originally introduced to `rosbag2_cpp` for the Reindexer class test. We then moved them to `rosbag2_tests`. I'm worried if they're exact copies of the originals, then they will also have a problem on OSX - we'll see when the above CI goes through. I'm ok with a little bit of the duplication right now  - the alternatives i see are moving this reindexer test to `rosbag2_tests`, or moving the test bag assets to `rosbag2_test_common` as a common resource for packages to use. I don't feel too particular about fixing that now though.

Note: there's a failing test problem right now that those will show for `rosbag2_py`'s Transport tests - see the conversation at https://github.com/ros2/rosbag2/pull/702. We'll need to make sure it's green before merging this, but I'll help push it through.

**Comment by jhdcs** [comment]
I thought I had copied over the ones you added. If there's a problem with the OSX test, then I'll delete them, and re-copy your database files again.

**Comment by jhdcs** [comment]
Looks like that might have been the old database! All right, swapping it out...

**Comment by emersonknapp** [comment]
Note: just waiting on https://github.com/ros2/rosbag2/pull/705 before moving this forward. I'll do a rebase and test asap (probably tomorrow morning after the WG meeting) then merge if it's green

**Comment by emersonknapp** [comment]
Gist: https://gist.githubusercontent.com/emersonknapp/a848456e0c1996c86fb34abc646482bc/raw/5eae459c861e4b2e910b1530f57b3ca917adc8c2/ros2.repos
BUILD args: --packages-up-to rosbag2_py ros2bag rosbag2_tests rosbag2
TEST args: --packages-select rosbag2_py ros2bag rosbag2_tests rosbag2
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/8087
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14203/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9001/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11895/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14306/)

**Comment by emersonknapp** [comment]
Windows warnings are the following two:
* https://github.com/ros2/rclpy/issues/752
* https://github.com/eclipse-cyclonedds/cyclonedds/issues/741

**Pull request merged**